docs/proposals/super-simple-discovery-api.md

```diff
@@ -0,0 +1,206 @@
+
+### Method: Bootstrap Token
+
+There won’t always be a trusted external endpoint to talk to and transmitting
+the locator file out of band is a pain.  However, we want something more secure
+than just hitting HTTP and trusting whatever we get back.  In this case, we
+assume we have the following:
+
+  * An address for at least one of the API servers (which will implement this API).
+    * This address is technically an HTTPS URL base but is often expressed as a bare domain or IP.
+  * A shared secret token
+
+An interesting aspect here is that this information is often easily obtained before the API server is configured or started.  This makes some cluster bring-up scenarios much easier.
+
+The user experience for joining a cluster would be something like:
+
+```
+kubeadm join --token=ae23dc.faddc87f5a5ab458 <address>
+```
+
+**Note:** This is logically a different use of the token from TLS bootstrap.  We harmonize these usages and allow the same token to play double duty.
+
+#### Implementation Flow
+
+`kubeadm` will implement the following flow:
+
+* `kubeadm` connects to the API server address specified over TLS.  As we don't yet have a root certificate to trust, this is an insecure connection and the server certificate is not validated.
+  * Implementation note: the API server doesn't have to expose a new and special insecure HTTP endpoint.
+* `kubeadm` requests a ConfigMap containing the kubeconfig file defined above.
+  * This ConfigMap exists at a well known URL: `https://<server>/api/v1/namespaces/kube-public/configmaps/cluster-info`
+  * This ConfigMap is really public.  Users don't need to authenticate to read this ConfigMap.  In fact, the client MUST NOT use a bearer token here as we don't trust this endpoint yet.
+* The API server returns the ConfigMap with the kubeconfig contents as normal
+  * Extra data items on that ConfigMap contains JWS signatures.  `kubeadm` finds the correct signature based on the `token-id` part of the token.  (Described below).
+* `kubeadm` verifies the JWS and can now trust the server.  Further
+  communication is simpler as the CA certificate in the kubeconfig file can be
+  trusted.
+
+
+#### NEW: Bootstrap Token Structure
+
+To first make this work, we put some structure into the token.  It has both a token identifier and the token value, separated by a dot.  Example:
+
+```
+ae23dc.faddc87f5a5ab458
+```
+
+The first part of the token is the `token-id`.  The second part is the `token-secret`.  By having a token identifier, we make it easier to specify *which* token you are talking about without sending the token itself in the clear.
+
+This new type of token is different from the current CSV token authenticator that is currently part of Kubernetes.  The CSV token authenticator requires an update on disk and a restart of the API server to update/delete tokens.  As we prove out this token mechanism we may wish to deprecate and eventually remove that mechanism.
```

Comment:
I'd say we need another design doc (or add to this one) for that stuff.  But my gut is that we would name it `system:bootstrap:<token-id>` and it would be a standard authenticator.  We'd look to deprecate the CSV file token authenticator as it isn't HA friendly and there is no easy way to expire or change tokens.